azurelinuxagent/common/event.py

```diff
@@ -548,11 +544,26 @@ def add_metric(self, category, counter, instance, value, log_event=False):
         event.parameters.append(TelemetryEventParam(GuestAgentPerfCounterEventsSchema.Value, float(value)))
         self.add_common_event_parameters(event, datetime.utcnow())
 
-        data = get_properties(event)
-        try:
-            self.save_event(json.dumps(data))
-        except EventError as e:
-            logger.periodic_error(logger.EVERY_FIFTEEN_MINUTES, "[PERIODIC] {0}".format(ustr(e)))
+        self.flush_or_save_event(event, message, immediate_flush)
+
+    def flush_or_save_event(self, event, message, immediate_flush):
+        """
+        Flush the event to wireserver if immediate_flush to set to true, else
+        save it disk if we fail to send or not required to flush immediately.
+        """
+        report_success = False
+        if immediate_flush and self.protocol is not None:
+            report_success = self.protocol.report_event([event])
+            if not report_success:
+                logger.error("Failed to send event: '{0}' directly to Wireserver. So, agent will save it to disk for periodic flush.", message)
```

Comment:
where do we delete the events when we fail to send them? important events should probably not be deleted, in case the send failure was transient and the next poll from the telemetry thread can send them

Reply:
yes, we save below #563 when we fail to send them